### PR TITLE
RATEST-154: Adding resource for clinical visit workflow

### PR DESCRIPTION
**Pull request by kdaud**
Ticket [ID](https://issues.openmrs.org/browse/RATEST-154)
Description -> Adding resource for clinical visit workflow

### CODE REVIEW

File: ui-tests/src/test/java/org/openmrs/reference/BookRequestAppointmentTest.java [hunk 1]
```diff
@@ -42,8 +42,8 @@ public void bookRequestAppointmentTest() throws Exception {
         ClinicianFacingPatientDashboardPage patientDashboardPage = activeVisitsPage.goToPatientDashboardOfLastActiveVisit();
         RequestAppointmentPage requestAppointmentPage = patientDashboardPage.clickOnRequest();
         requestAppointmentPage.enterAppointmentType("Oncology");
-        requestAppointmentPage.enterValue("0");
-        requestAppointmentPage.selectUnits("Day(s)");
+        requestAppointmentPage.enterMinimumValue("0");
+        requestAppointmentPage.selectMinimumUnits("Day(s)");
```
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/BookRequestAppointmentTest.java:46]: Refactored the function names [h](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/9c172b398f03b79b816894c176df11643432a6c2/ui-tests/src/test/java/org/openmrs/reference/page/RequestAppointmentPage.java#L31)er[e](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/9c172b398f03b79b816894c176df11643432a6c2/ui-tests/src/test/java/org/openmrs/reference/page/RequestAppointmentPage.java#L35) for better java naming convention

File: ui-tests/src/test/java/org/openmrs/reference/DeleteRequestAppointmentTest.java [hunk 2]
```diff
@@ -41,8 +41,8 @@ public void deleteRequestAppointmentTest() {
         ClinicianFacingPatientDashboardPage patientDashboardPage = activeVisitsPage.goToPatientDashboardOfLastActiveVisit();
         RequestAppointmentPage requestAppointmentPage = patientDashboardPage.clickOnRequest();
         requestAppointmentPage.enterAppointmentType(SERVICE_NAME);
-        requestAppointmentPage.enterValue("0");
-        requestAppointmentPage.selectUnits("Day(s)");
+        requestAppointmentPage.enterMinimumValue("0");
+        requestAppointmentPage.selectMinimumUnits("Day(s)");
```
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/DeleteRequestAppointmentTest.java:45]: Refer to the comment [here](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/502/commits/9c172b398f03b79b816894c176df11643432a6c2#r668817026)

Comment (kaweesi) [reply ui-tests/src/test/java/org/openmrs/reference/DeleteRequestAppointmentTest.java:45]: any reasons for renaming this?

Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/DeleteRequestAppointmentTest.java:45]: > any reasons for renaming this?

The reason for renaming is [here](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/502#discussion_r668817026)

File: ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java [hunk 3]
```diff
@@ -78,6 +78,8 @@
     private static final By ACTIVE_VISIT_MESSAGE = By.cssSelector("active-visit-message");
     private static final By ALLERGIES_LINK = By.id("allergyui-editAllergies");
     private static final By CONDITIONS_LINK = By.cssSelector(".conditions .info-header i.right");
+    private static final By ATTACHMENTS_LINK = By.cssSelector("a[id='attachments.attachments.visitActions.default']");
+    private static final By ACTIVE_VISIT = By.cssSelector("a.ng-binding");
```
Comment (kaweesi) [inline ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java:82]: this css needs to be more improved, work with better element lookup, consider parent elements to restrict what can be retrieved here

File: ui-tests/src/test/java/org/openmrs/reference/page/RequestAppointmentPage.java [hunk 4]
```diff
@@ -21,19 +28,36 @@ public void enterAppointmentType(String type) {
...
+        setText(MAX_FRAME_VALUE, value);
+    }
+    
+    public void selectMaximumUnits(String units) {
+        selectFrom(MAX_FRAME_UNITS, units);
+    }
+    
+    public void addAppointmentRequestNote(String note) {
+        findElement(APPOINTMENT_REQUEST_NOTE).clear();
+        setText(APPOINTMENT_REQUEST_NOTE, note);
```
Comment (kaweesi) [inline ui-tests/src/test/java/org/openmrs/reference/page/RequestAppointmentPage.java:49]: setText should automatically overwrite the existing value

File: ui-tests/src/test/java/org/openmrs/reference/page/PatientVisitsDashboardPage.java [hunk 5]
```diff
@@ -89,6 +90,11 @@ public MergeVisitsPage clickOnMergeVisits() {
         return new MergeVisitsPage(this);
     }
 
+    public ClinicianFacingPatientDashboardPage goToPatientDashboard() {
+        clickOn(RETURN_TO_DASHBOARD);
```
Comment (kaweesi) [inline ui-tests/src/test/java/org/openmrs/reference/page/PatientVisitsDashboardPage.java:94]: You can refactor https://github.com/openmrs/openmrs-distro-referenceapplication/blob/master/ui-tests/src/test/java/org/openmrs/reference/page/StylesGuidePage.java#L14 into the parent class and use that there and here

File: ui-tests/src/test/java/org/openmrs/reference/page/RequestAppointmentPage.java (hunk 4 above)
Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/page/RequestAppointmentPage.java:49]: > setText should automatically overwrite the existing value

Does this mean eliminating findElement().clear(); completely? and every where I had applied it as well

File: ui-tests/src/test/java/org/openmrs/reference/page/ConditionsPage.java [hunk 6]
```diff
@@ -14,6 +17,7 @@
     private static final By FIRST_CONDITION_NAME = By.xpath("//table/tbody[2]/tr[1]/td[1]");
     private static final By EDIT = By.cssSelector("i[title='Edit Condition: ']");
     private static final By DELETE = By.cssSelector("i[title='Delete']");
+    private static final By CONDITIONS_LIST = By.cssSelector("tr.clickable-tr");
```
Comment (kaweesi) [inline ui-tests/src/test/java/org/openmrs/reference/page/ConditionsPage.java:20]: this list should also be referenced with a more reliable referencing mechanism. consider parent elements that would qualify any sub elements

File: ui-tests/src/test/java/org/openmrs/reference/page/AddOrEditAllergyPage.java [hunk 7]
```diff
@@ -23,6 +26,9 @@
     private static By DRUG_ID;
     private static String REACTION_XPATH = "//div[@id='reactions']/ul/li/label[text()=\"";
     private static By REACTION_ID;
+    private static By ALLERGY_NOTE = By.id("allergy-comment");
+    private static By ALLERGIES_LIST = By.cssSelector("#allergies tbody tr td.allergen");
+    private static By RETURN = By.cssSelector("#content > button.cancel");
```
Comment (kaweesi) [inline ui-tests/src/test/java/org/openmrs/reference/page/AddOrEditAllergyPage.java:31]: this also needs to be more qualified such addition of any extra cancel buttons won't break the test

**Pull request closed without merging**